src/fsharp/service/SemanticClassificationKey.fs

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace FSharp.Compiler.SourceCodeServices
+
+open System
+open System.IO
+open System.IO.MemoryMappedFiles
+open System.Reflection.Metadata
+open System.Runtime.Serialization.Formatters.Binary;
+
+open FSharp.NativeInterop
+open FSharp.Compiler.Text
+open FSharp.Compiler.Text.Pos
+open FSharp.Compiler.Text.Range
+
+#nowarn "9"
+
+[<Sealed>]
+type SemanticClassificationKeyStoreView(mmf: MemoryMappedFile, length) =
+    member _.ReadRange(reader: byref<BlobReader>) =
+        let startLine = reader.ReadInt32()
+        let startColumn = reader.ReadInt32()
+        let endLine = reader.ReadInt32()
+        let endColumn = reader.ReadInt32()
+        let fileIndex = reader.ReadInt32()
+
+        let posStart = mkPos startLine startColumn
+        let posEnd = mkPos endLine endColumn
+        mkFileIndexRange fileIndex posStart posEnd
+
+    member this.ForEach() =
+        use view = mmf.CreateViewAccessor(0L, length)
+        let mutable reader = BlobReader(view.SafeMemoryMappedViewHandle.DangerousGetHandle() |> NativePtr.ofNativeInt, int length)
+
+        reader.Offset <- 0
+        seq {
+            while reader.Offset < reader.Length do
+                let m = this.ReadRange &reader
+                let sct = reader.ReadInt32()
+                yield struct(m, (enum<SemanticClassificationType>(sct)))
+        }
+
+and [<Sealed>] SemanticClassificationKeyStore(mmf: MemoryMappedFile, length) =
+    let mutable isDisposed = false
+    let checkDispose() =
+        if isDisposed then
+            raise (ObjectDisposedException("SemanticClassificationKeyStore"))
+
+    member _.GetView() =
+        checkDispose()
+        SemanticClassificationKeyStoreView(mmf, length)
+
+    interface IDisposable with
+
+            member _.Dispose() =
+                isDisposed <- true
+                mmf.Dispose()
+
+and [<Sealed>] SemanticClassificationKeyStoreBuilder() =
+
+    let b = BlobBuilder()
+
+    member _.WriteAll (semanticClassification: struct(range * SemanticClassificationType)[]) =
+        let formatter = BinaryFormatter()
```

Comment:
I don't think we need the `BinaryFormatter` or the `MemoryStream`.

You should just be able to get the raw pointer of the array and call `WriteBytes` with the pointer and the byte count: https://docs.microsoft.com/en-us/dotnet/api/system.reflection.metadata.blobbuilder.writebytes?view=net-5.0#System_Reflection_Metadata_BlobBuilder_WriteBytes_System_Byte__System_Int32_

You will have to pin the array, I think you can do:
```fsharp
use ptr = fixed semanticClassification
b.WriteBytes(ptr, semanticClassification.Length)
```

Reply:
ah, okay. Thanks for the references. I'll try to use them.

Reply:
This doesn't quite typecheck. 
`ptr` is of type `nativeptr<struct(range * SemanticClassificationType)>` but `NativePtr.ofNativeInt` expects a `nativeint`.

I guess you meant `toNativeInt` rather than `ofNativeInt` but even that would give me an `nativeint` rather than `byte`. 
I thought I'll be able to just cast it to `byte` using `byte(NativePtr.toNativeInt ptr)` but I get the following error:

```
FS0001: a generic construct requires that the type 'struct(range * SemanticClassificationType)'  is an unmanaged type
```

Reply:
Ah, that's right. F# doesn't support unmanaged generic constructed types yet, though C# does. One way to handle this is to create a separate struct type that contains the `range` and `SemanticClassificationType` instead of using a struct tuple.

Something like this:
```
[<Struct>]
type FSharpSemanticClassificationItem =
    val Range: range
    val Type: SemanticClassificationType
    new(range, ty) = { Range = range; Type = ty }
```
Then the type of the array becomes `FSharpSemanticClassificationItem []`.